code/service/src/main/scala/com/namely/chiefofstate/readside/ReadSideHandler.scala

```diff
@@ -103,27 +105,73 @@ private[readside] class ReadSideHandlerImpl(
   }
 }
 
-/**
- * Processes events read from the Journal
- *
- * @param event          the actual event
- * @param eventTag       the event tag
- * @param resultingState the resulting state of the applied event
- * @param meta           the additional meta data
- * @return an eventual HandleReadSideResponse
- */
 private[readside] trait ReadSideHandler {
 
+  def onBeginProcess(): Unit = {}
+  def onEndProcess(): Unit = {}
+
   /**
-   * handles a read side message
+   * Processes events read from the Journal.
+   * Exponentially backoff with a 10& gain modifier until it reaches the upper threshold of maxBackoffSeconds.
+   * If maxAttempts is set to a value 0 or less, it will backoff indefinitely, otherwise it
+   * will run a number of times up to to the maxAttempts.
    *
-   * @param event
-   * @param eventTag
-   * @param resultingState
-   * @param meta
-   * @return
+   * @param event               the actual event
+   * @param eventTag            the event tag
+   * @param resultingState      the resulting state of the applied event
+   * @param meta                the additional meta data
+   * @param maxAttempts         the max number of attempts before quit, infinite if set to 0
+   * @param minBackoffSeconds   the minimum number of seconds to backoff
+   * @param maxBackoffSeconds   the maximum number of seconds to backoff
+   * @return                    Boolean for success
    */
   def processEvent(
+      event: com.google.protobuf.any.Any,
+      eventTag: String,
+      resultingState: com.google.protobuf.any.Any,
+      meta: MetaData,
+      maxAttempts: Int = 0,
+      minBackoffSeconds: Long = 1,
+      maxBackoffSeconds: Long = 30): Boolean = {
+
+    // Recursive function that incorporates exponential backOff
+    @tailrec
+    def loop(numAttempts: Int = 0): Boolean = {
+      val isSuccess: Boolean = doProcessEvent(event, eventTag, resultingState, meta)
+
+      if (!isSuccess && (maxAttempts <= 0 || numAttempts >= maxAttempts)) {
+        val backoffSeconds: Long = Math.min(maxBackoffSeconds, (minBackoffSeconds * Math.pow(1.1, numAttempts)).toLong)
+
+        Thread.sleep(Duration.ofSeconds(backoffSeconds).toMillis)
```

Comment:
This is a bit dangerous because we are inside an actor. The projection is an actor making the grpc call. Let us just becareful here

Reply:
What do you suggest in this scenario?

Reply:
I think we should make the grpc call asynchronous and make use scala retry on future. That will make it more resilient.
For instance you can use these libraries:

- https://github.com/softwaremill/retry (recommended because of the implementation is based upon this algo https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/)
- https://doc.akka.io/docs/akka/current/futures.html#retry: the downside with this is you need the actor system and I don't think it is necessary.

Reply:
So instead of using the blocking stub of the grpc client we can use the non-blocking one.

Reply:
@zenyui we need to take into consideration this:
for the projection to listen and read events persisted the sharded daemon is hooked it into the existing actor system. With that type of setup I have some worries around the execution context. Since the actor system is making use of the default dispatcher the retry mechanism can hinder a bit the performance of the system(writeside). I think we may have to push this feature into its own execution context to avoid hindering the whole actor system.

Reply:
The reason I am saying this is assuming all the ten actors will have to do the retry.

Reply:
I'm fine with using the retry package, however it's using promises, which is essentially blocking an entire thread just like  a Thread.sleep. Why would this be superior aside from their jitter implmentation?

Reply:
Funny thing is... I started with the amazon algo, then was asked to simplify it :D

Reply:
That is why I am saying we may have to consider a separate thread executor. Whatever we do needs another thread executor to think of it properly. These libraries are just suggestions.

Reply:
The projection JDBC handler we are using does not have its own its thread pool compared to the slick version we stopped using.

Reply:
@Tochemey the call is already blocking, I don't see the issue. I don't mind creating a new thread pool for remote calls, but using a future won't change anything. Under the hood, they are likely just using sleep.